hybridse/src/udf/default_defs/window_functions_def.cc

```diff
@@ -90,6 +92,131 @@ node::ExprNode* BuildAt(UdfResolveContext* ctx, ExprNode* input, ExprNode* idx,
     return res;
 }
 
+
+template <typename T>
+struct RingBuffer {
+    explicit RingBuffer(size_t sz, const T& default_val) : idx(0), sz(sz), data(sz, default_val) {}
+
+    template <typename... Arg>
+    void Emplace(Arg&&... val) {
+        data[idx % sz] = T(std::forward<Arg>(val)...);
+        idx++;
```

Comment:
- the position of 0 or nth is not the real nth?  the position like first or last is need to record
- consider the case that n is much larger than the number of records(
extreme case or incorrect input).  it is not recommended to allocate the `n` size buffer at the begging

Reply:
if use vector, one optional solution is keep nth as the member. if the size of vector is less than nth, push_back directly. otherwise compute the real idx as current implement. Absolutely, it need additional judgement  in `At` or `output` function if the size less than nth

Reply:
Dynamically sized vector issues many relocation if nth is huge. In that case, deque is better.

Reply:
using deque has disadvantage too, according https://en.cppreference.com/w/cpp/container/deque
> deques typically have large minimal memory cost; a deque holding just one element has to allocate its full internal array (e.g. 8 times the object size on 64-bit libstdc++; 16 times the object size or 4096 bytes, whichever is larger, on 64-bit libc++)

Reply:
I choose the `std::queue<Container = std::list>`, which is out of relocation , not so huge startup space, and *O(1)*  push and pop. General memory space should a little higher compared to `std::vector`. 

Comment below if there is a better container.